### DISCUSSION TITLE
Ambiguous legacy escape codes

### DISCUSSION THREAD

**Discussion by bound-variable** [kovidgoyal/kitty#4778] [Q&A] [answered]
In the [documentation](https://sw.kovidgoyal.net/kitty/keyboard-protocol/#legacy-text-keys) it says,

> Many of the legacy escape codes are ambiguous with multiple different key presses yielding the same escape code(s), for example, ctrl+i is the same as tab, ctrl+m is the same as Enter, ctrl+r is the same ctrl+shift+r, **etc**. To resolve these use the [disambiguate progressive enhancement](https://sw.kovidgoyal.net/kitty/keyboard-protocol/#disambiguate).

What does **et cetera** refer to here? What other ambiguous legacy escape codes are there? I'd like to know because I removed the keybinding `Ctrl-M` in zsh (via `bindkey -r ^M`) and my `Enter` key stopped working. It would be nice to know ahead of time what other keys are liable to thus break.

I found that I can compare the bytes under **[Legacy ctrl mapping of ASCII keys](https://sw.kovidgoyal.net/kitty/keyboard-protocol/#legacy-ctrl-mapping-of-ascii-keys)** with the CSI under **[Functional key definitions](https://sw.kovidgoyal.net/kitty/keyboard-protocol/#functional-key-definitions)**. For example, `Ctrl+I` has byte 9, and `Tab` has CSI 9. So this is one way of identifying mappings? If so, I don't understand why `Tab` continues to work if I unbind `Ctrl+I` in zsh, but `Enter` stops working if I unbind `Ctrl+M` (`Enter` and `Ctrl+M` both have the number 13).

**Comment by kovidgoyal** [comment]
These are mappings in zsh, not kitty. You should ask in the zsh project.

**Reply by bound-variable** [reply]
> These are mappings in zsh, not kitty. You should ask in the zsh project.

But I quoted kitty's documentation. So, I'm confused.

On my system, fzf binds `Ctrl-i` to the command `fzf-completion` in zsh. I asked on Reddit what this command did because pressing `Ctrl-i` didn't do anything for me. The answer I received was that terminals usually bind the `Tab` key to `Ctrl-i`. Then I learned that there are other such bindings, e.g., `Ctrl-m` with `Enter`.

So these bindings aren't made from kitty?

I want to change the default bindings in zsh, but when I do so, certain keys like `Enter` stop working even though I didn't explicitly unbind `Enter`, I unbound `Ctrl-m`.  I'm trying to figure out how to avoid these unintended consequences.

**Comment by kovidgoyal** [comment]
What is it you are trying to do? A ambiguous legacy key mapping means more than one key press produces the same bytes. There are a multitude of such mappings in the legacy terminal key encoding. Removing them in zsh doesnt achieve anything, other than making all the keys that produce those bytes unusable in zsh.

**Reply by bound-variable** [reply]
I'm trying to remove bindings in zsh that I will never use, while also creating bindings that I will use. The problem is that removing certain bindings, such as `Ctrl-m`, has unintended consequences, such as losing `Enter`.

For example, I want `Enter`, but I don't want `Ctrl-m`. I want `Tab`, but I don't want `Ctrl-i`. Even if that's not possible, I would still like to know about such mappings.

**Comment by trygveaa** [comment]
> In the [documentation](https://sw.kovidgoyal.net/kitty/keyboard-protocol/#legacy-text-keys) it says,
> 
> > Many of the legacy escape codes are ambiguous with multiple different key presses yielding the same escape code(s), for example, ctrl+i is the same as tab, ctrl+m is the same as Enter, ctrl+r is the same ctrl+shift+r, **etc**. To resolve these use the [disambiguate progressive enhancement](https://sw.kovidgoyal.net/kitty/keyboard-protocol/#disambiguate).
> 
> What does **et cetera** refer to here? What other ambiguous legacy escape codes are there? I'd like to know because I removed the keybinding `Ctrl-M` in zsh (via `bindkey -r ^M`) and my `Enter` key stopped working. It would be nice to know ahead of time what other keys are liable to thus break.

Don't know if this is all, but in addition to Ctrl-M/Enter and Ctrl-I/Tab, you have Ctrl-[ for Escape, Ctrl-? for Backspace and Ctrl-H for Ctrl-Backspace. You can check if a key uses Ctrl-<character> by running `showkey -a` and pressing the key. If it appears just as `^` plus one character, it means Ctrl and that character.

> I found that I can compare the bytes under **[Legacy ctrl mapping of ASCII keys](https://sw.kovidgoyal.net/kitty/keyboard-protocol/#legacy-ctrl-mapping-of-ascii-keys)** with the CSI under **[Functional key definitions](https://sw.kovidgoyal.net/kitty/keyboard-protocol/#functional-key-definitions)**. For example, `Ctrl+I` has byte 9, and `Tab` has CSI 9. So this is one way of identifying mappings?

Not sure if that's reliable, it only seems to match for escape, enter and tab.

> If so, I don't understand why `Tab` continues to work if I unbind `Ctrl+I` in zsh, but `Enter` stops working if I unbind `Ctrl+M` (`Enter` and `Ctrl+M` both have the number 13).

Are you sure you unbound it correctly? If I unbind Ctrl-I in zsh, tab stops working. Note that you need to place the Ctrl-I character inside quotes for the command to work (and you need to insert the actual character in the command, which you can do by pressing Ctrl-V Ctrl-I).

**Reply by bound-variable** [reply]
`showkey` is great, thanks.

Yes, something strange is happening with `Ctrl+I`. It's not printed when I press `Tab` in `showkey`, but `Ctrl+M` and `Ctrl+?` are printed when pressing `Enter` and `Backspace`, respectively. My guess is this is because fzf is doing something with it.

Anyhow, I think I can figure this out with `showkey` now, thanks again.

**Comment by bound-variable** [comment, marked as answer]
I found [this gist](https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#general-ascii-codes)  on **general ascii escape codes**, which, to my green eyes, seems to cover everything I care about.
